### PR TITLE
Hash exist url

### PR DESCRIPTION
**Pull request by tcitworld**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | yes
| Documentation | no
| Translation   | no
| Fixed tickets | #3157 
| License       | MIT

### CODE REVIEW

**Comment by tcitworld** [comment]
Godammit Travis.

**Comment by j0k3r** [comment]
Rebase needed

**Comment by nicosomb** [comment]
Just FYI, this feature will also be useful to add an index on entries. 
It's complicated to add an index on URL (because URL can have so many letters), but it will be easier with hashes.

**Comment by shtrom** [comment]
Hey there,

Not sure if I'm reading this correctly, but does this mean that the client is now in charge of the hashing before hitting the URL? This seems to introduce an unnecessary coupling, where client and API now need to agree on the hash to use. It would seem more flexible to not change the API, and do the hashing entirely in the backend, just before hitting the DB.

Speaking of hash agility, I wonder if SHA-512 is not a bit overkill here. Noting that this is not a security feature, I think we could just use MD5, which would be faster. The only drawback might be the risk of increase collision, but I guess this could be handled with searching by hash, then confirming a hit by doing a full-text comparison. The hash would however not be usable as an primary key, but I don't think that would be an issue.

Better yet, with all the hashing in the backend, what algorithm to use could be a simple configuration option!

**Comment by Kdecherf** [comment]
> Not sure if I'm reading this
> [wallabag/src/Wallabag/ApiBundle/Controller/EntryRestController.php](https://github.com/wallabag/wallabag/blob/642528bb15293405ccf0160f06aff342f165714f/src/Wallabag/ApiBundle/Controller/EntryRestController.php#L24-L25)
> 
> Lines 24 to 25 in [642528b](/wallabag/wallabag/commit/642528bb15293405ccf0160f06aff342f165714f)
> 
>       *          {"name"="hashedurl", "dataType"="string", "required"=true, "format"="An url", "description"="SHA512 Url to check if it exists"}, 
>       *          {"name"="hashedurls", "dataType"="string", "required"=false, "format"="An array of urls (?urls[]=http...&urls[]=http...)", "description"="SHA512 Urls (as an array) to check if it exists"} 
> 
> correctly, but does this mean that the client is now in charge of the hashing before hitting the URL? This seems to introduce an unnecessary coupling, where client and API now need to agree on the hash to use. It would seem more flexible to not change the API, and do the hashing entirely in the backend, just before hitting the DB.

The main concern in the linked issue is a privacy one. The idea is to provide a way for the client to quickly check if a given address exists in the database without leaking the actual searched address in the webservers access logs. In this case the hash must be done on the client side.

> Better yet, with all the hashing in the backend, what algorithm to use could be a simple configuration option!

The hashing algorithm may request a database change to update the length of the corresponding field in the table: we should always provide a fixed char field for better indexation.

**Comment by j0k3r** [comment]
Good point @Kdecherf. I was _all-in_ with @shtrom until I re-read the privacy issue.
I also think the algorithm won't be a parameter, it'll be hardcoded to md5.

**Comment by Kdecherf** [comment]
> I also think the algorithm won't be a parameter, it'll be hardcoded to md5.

Maybe we could at least use sha1?

**Comment by j0k3r** [comment]
> Maybe we could at least use sha1?

Okay

**Comment by shtrom** [comment]
> Good point @Kdecherf. I was _all-in_ with @shtrom until I re-read the privacy issue.
> I also think the algorithm won't be a parameter, it'll be hardcoded to md5.

Ah, yeah, I missed the privacy point. Good one, yeah!

tl;dr: ignore my previous comment, it's mainly invalid if you have privacy in
mind.

If we have some security/privacy concerns, then I think my argument about MD5
is a bit moot. We should probably use a more secure algorithm so it's
difficult to reverse the hash (which is not that much with MD5).

In addition, I think I saw a mention of a salt by username somewhere; I think
it would be a good idea.

On related matters, did you see how Troy Hunt implemented k-anonimyt for HaveIBeenPwend? https://www.troyhunt.com/were-baking-have-i-been-pwned-into-firefox-and-1password/#enablinganonymoussearcheswithkanonymity

**Comment by j0k3r** [comment]
While it might eventually be a good idea to enable salt I think it’s really overkill for the feature.
I do not think it’s important to preserve that much the url we want to check. We are not dealing with password, email or a real sensible information here FMPOV.
